src/cluster/src/cli/error.rs

```diff
@@ -0,0 +1,68 @@
+use std::io::Error as IoError;
+
+use fluvio::FluvioError;
+use fluvio_extension_common::output::OutputError;
+use fluvio_extension_common::target::TargetError;
+use fluvio_runner_local::RunnerError;
+use crate::{ClusterError, CheckError};
+
+/// Cluster Command Error
+#[derive(thiserror::Error, Debug)]
+pub enum ClusterCliError {
+    #[error(transparent)]
+    IoError(#[from] IoError),
+    #[error("Output Error")]
+    OutputError(#[from] OutputError),
+    #[error("Target Error")]
+    TargetError(#[from] TargetError),
+    #[error("Fluvio cluster error")]
+    ClusterError(#[from] ClusterError),
+    #[error("Fluvio cluster pre install check error")]
+    CheckError(#[from] CheckError),
+    #[error("Fluvio client error")]
+    ClientError(#[from] FluvioError),
+    #[error("Runner error")]
+    RunnerError(#[from] RunnerError),
+    #[error("Unknown error: {0}")]
+    Other(String),
+}
+
+impl ClusterCliError {
+    pub fn into_report(self) -> color_eyre::Report {
+        use color_eyre::Report;
+
+        match self {
+            Self::ClusterError(cluster) => cluster.into_report(),
+            _ => Report::from(self),
+        }
+    }
+}
+
+// This impl is here so that it is only compiled under "cli" feature flag
+impl ClusterError {
+    pub fn into_report(self) -> color_eyre::Report {
+        use color_eyre::{Report, Section};
+        use fluvio_helm::HelmError as Helm;
+        use CheckError as Check;
+        use crate::K8InstallError::*;
+
+        match &self {
+            Self::InstallK8(PreCheck(Check::HelmError(Helm::FailedToConnect))) => {
+                let report = Report::from(self);
```

Comment:
Isn't this already same suggestion done in Precheck? seems like this is duplicate logic

Reply:
I suggest add `Suggestion` similar to `Display` of `Debug` trait.   This allow more flexibility and suggestion as well as push logic to near the error.  For example:

```
pub trait Suggestion {

       fn format(&self,  report: Report);
}
```

then you can add them to Helm error
```
impl Suggestion for HelmError {

     .....
}
```

Reply:
I don't think this is duplicate logic. In `fluvio-helm`, we have this from errors.rs:

```rust
pub enum HelmError {
...
    #[error("Failed to connect to Kubernetes")]
    FailedToConnect,
...
}
```

and here we are adding a suggestion of how to possibly remedy it.

I don't really understand this `Suggestion` trait you are recommending. The current suggestions are already extension methods on the `eyre::Report` type via the `color_eyre::Section` trait. I actually think that the way things are laid out here works out pretty well. Here are the considerations that were in my head when I wrote it this way:

- Error types belonging to _library crates_ (e.g. HelmError and CheckError) should be strictly data structures with Display/Debug implementations. Some of these error data structures are nested, i.e. they hold another error inside them, which indicates that the present-level error was "caused by" the inner error. This nesting of error types creates a context that describes a relationship between the high-level errors which were caused by subsequent low-level errors.
- Different _application crates_ can define more specialized ways of rendering errors received from libraries. This is the role of the `eyre`/`color_eyre` crate and of the `Report` type from them. A Report simply wraps a library's error data type and uses the information in the Display/Debug implementations to produce a nicer, CLI-specialized rendering of those errors. It also allows you to add special "sections" to the "Report", (this is where the current .`suggestion()` lines come into play), which are all collected together and printed in a nice colorized way to the terminal (see https://docs.rs/color-eyre/0.5.8/color_eyre/section/trait.Section.html).
  - This way, different types of applications with different error-rendering needs do not need to pull in specialized code for rendering errors for a different application type. For example, a web service using `fluvio-cluster` should not need to link to the code which renders `fluvio-cluster` errors for the CLI. Instead, it should provide its own application-specific rendering, such as mapping the error into an HTTP resposne with JSON. Similarly, this HTTP-JSON error-rendering would not be linked into the CLI binary target.
  - This `into_report` function in `src/cluster/src/cli/error.rs` was where I thought there was the best boundary between the `fluvio-cluster` _library errors_ and the _application error rendering_. One thing to note is that this is a unique task and that the `into_report` function provides a single home for us to add more CLI-specific annotations to our application errors, and they won't spill out into the rest of the codebase.

Let me know if you have any problems with this design, but I don't see any glaring problems that will harm the codebase in the long run. We can always tweak and update the exact layout later if we find pain points.

Reply:
Initial thought is that  `Suggestion` trait might be more appropriate for Fluvio.   However, if `Report` type is sufficient for now, we can defer that idea later. 

Still, `into_report` should be push down to individual Error or at least to top level Error in the individual extension.  Implementation at CLI creates explicit coupling between them.   For example, if we run `Cluster` extension by itself, it will have different output.   CLI is not expected to perform any post processing.  It's job is to hand off argument to extension

Reply:
> to top level Error in the individual extension

As of right now, the only suggestions that exist are in `fluivo-cluster` for the `ClusterError` type. Other `into_report` implementations will live in the respective crate for other extensions.

Reply:
Yes, I see that now.   It looks good now